Submit Analytics Data (EXPOSUREAPP-4823)

kolyaopahle opened the pull request:
### Description
This PR contains the backend logic for Analytics submission
It should only be merged after all other Analytics PRs are merged as it will be used to combine front and backend interactions.

### Testing
There are quite a few options in the Data Donation tab of the Tester Menu to already test functionality, but there is currently no working testing environment providing the necessary backend.
You could use the cwa-app-cli project which has support for the backend or check logcat for errors

kolyaopahle committed:
initial commit of Analytics upload code (lots of missing pieces)
kind of working: Definition of backends and endpoints, donor data collection (no donors implemented here)
probably broken: SafetyNet attestation, actual server communication

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'feature/4321-ppa-main' into feature/4823-submit-analytics-data

kolyaopahle committed:
added checks for analytics submission abortion

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
added analytics submission periodic worker definition

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
lint fixes

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Fixed kotlin support for IntoSet injection
Added AnalyticsModule to DataDonationModule
Added WorkerBinder definition for DataDonationAnalyticsPeriodicWorker
Added DataDonationAnalyticsPeriodicWorker scheduling to CoronaWarnApplication
Added missing mocks to WorkerBinderTest

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'feature/4321-ppa-main' into feature/4823-submit-analytics-data

kolyaopahle committed:
Fixed injectData call in ExposureRiskMetadataDonor
Added more logging to Analytics
Implemented first AnalyticsTests
Fixed auto formatted removal of io.kotest.matchers.collections.shouldContainAll in WorkerBinderTest (curse you Optimize imports)

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Implemented Test for ExposureRiskMetadataDonor

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Added check for isEnabled during Analytics submission
Added DataDonationAnalyticsOneTimeWorker to allow for random distribution of Analytics submission events

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Moved contribution collection to own function to help with test menu features

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'feature/4321-ppa-main' into feature/4823-submit-analytics-data

kolyaopahle committed:
Added analytics testing features

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
small fixes to the WorkerBinderTest
added missing worker binding to WorkerBinder

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'feature/4321-ppa-main' into feature/4823-submit-analytics-data

d4rken committed:
Put Analytics test options into an extra card.

kolyaopahle committed:
in my defense, it was quite late

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge remote-tracking branch 'origin/feature/4823-submit-analytics-data' into feature/4823-submit-analytics-data

kolyaopahle committed:
Merge branch 'feature/4321-ppa-main' into feature/4823-submit-analytics-data

kolyaopahle committed:
Moved AnalyticsSettings to own package
Updated the @IntoSet providers in AnalyticsModule to match the actual DonorModules

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Added analyticsEnabled preference to AnalyticsSettings

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Added a deleteData function to the DonorModule interface which can be used to remove persisted data if a user revokes analytics consent

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Added skeleton to persist last analytics submission on tester devices
Currently the tester version also does not store the data

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Fixed borked unit tests

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Implemented File based storage of Last submitted analytics file on tester devices

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'feature/4321-ppa-main' into feature/4823-submit-analytics-data

kolyaopahle committed:
Added calls to enable and disable analytics to the OnboardingAnalyticsViewModel

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Fixed OnboardingAnalyticsFragmentTest

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Turns out AnalyticsConfigMapping was not implemented yet, fixed that

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Added a timestamp to the stored Analytics Submission

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Fixed unit tests in AnalyticsTest

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Updated AnalyticsConfigMapper to throw an exception if app config is invalid

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

d4rken commented on Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt:
```diff
@@ -0,0 +1,87 @@
+    @BaseGson private val baseGson: Gson
+) : LastAnalyticsSubmissionLogger {
+    private val analyticsDir = File(context.cacheDir, "analytics_storage")
+    private val analyticsFile = File(analyticsDir, "analytics.bin")
```
Minor, `analytics.bin` seems very generic, isn't it `last_analytics.bin`?

d4rken commented on Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt:
```diff
@@ -0,0 +1,87 @@
+
+    override suspend fun getLastAnalyticsData(): LastAnalyticsSubmission? = withContext(dispatcherProvider.IO) {
+        try {
+            gson.fromJson(analyticsFile)
```
Based on issues we had with the exposure detection which also stores json to file:
We should prevent empty files from causing invalid data classes being created by GSON (examples in `ExposureDetectionTrackerStorageTest`).

Despite not being allowed GSON can create data classes with null fields that shouldn't be null (unsafe init), which the JVM will only notice once accessed.

TL;DR:  `.also { requireNotNull(it.timestamp) }`

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt:
```diff
@@ -4,20 +4,59 @@ import dagger.Reusable
+            !rawConfig.privacyPreservingAnalyticsParameters.hasCommon() &&
+            !rawConfig.privacyPreservingAnalyticsParameters.hasPpac()
+        ) {
+            throw ApplicationConfigurationInvalidException(
```
Is the expected behavior to throw here (and discard the whole config), or would it be okay to return a default of "analytics disabled", as analytics being disabled would still allow the rest of the app to function?

kolyaopahle replied:
i talked to @mlenkeit about this, he mentioned that the analytics config part of the app config should not be optional so we can assume a broken config if no analytics config is provided

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt:
```diff
@@ -4,20 +4,59 @@ import dagger.Reusable
-        override val probabilityToSubmit: Float = 1f,
-        override val probabilityToSubmitAfterRiskCalculation: Float = 1f,
-        override val probabilityToSubmitNewExposureWindows: Float = 1f
+        override val probabilityToSubmit: Double = 1.0,
```
Should the defaults be 1.0, or would it be better if the defaults were 0.0, i.e. disabled? Does the spec say?

kolyaopahle replied:
not really, i'll remove the defaults for now as there is no way to instantiate the defaults anyways at the moment

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:
```diff
@@ -1,34 +1,156 @@
+            )
 
-    val isEnabled: Boolean = true
+            val attestation = deviceAttestation.attest(ppaAttestationRequest)
```
This only means that SafetyNet passed "generally". The analytics config contains specific requirements that you need to check, i.e.

`attestation.requirePass(analyticsConfig.requirements)`

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:
```diff
@@ -1,34 +1,156 @@
+                .setPayload(ppaData)
+                .build()
+
+            dataDonationAnalyticsServer.uploadAnalyticsData(ppaContainer)
```
I'd :heart: a log statement before attestation, after attestation pass and after upload.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:
```diff
@@ -1,34 +1,156 @@
+
+        if (success) {
+            settings.lastSubmittedTimestamp.update {
+                Instant.now()
```
`Timestamper` would like to be part of the analytics gang.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:
```diff
@@ -1,34 +1,156 @@
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    suspend fun stopDueToProbabilityToSubmit(): Boolean {
+        val submitRoll = Random.nextDouble(0.0, 1.0)
+        return submitRoll > appConfigProvider.getAppConfig().analytics.probabilityToSubmit
```
`>=`?

What if we have chance 1.0 and roll a 1.0?

kolyaopahle replied:
we can't roll a 1.0 afaik `nextDouble` is (inclusive, exclusive)

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:
```diff
@@ -1,34 +1,156 @@
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToLastSubmittedTimestamp(): Boolean {
+        val lastSubmit = settings.lastSubmittedTimestamp.value ?: return false
+        return lastSubmit.plus(Hours.hours(LAST_SUBMISSION_MIN_AGE_HOURS).toStandardDuration()).isAfterNow
```
`isAfterNow` internally calls `Instant.now` right? `TimeStamper` so we can fake this in tests?

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:
```diff
@@ -1,34 +1,156 @@
+        return onboarding.plus(Hours.hours(ONBOARDING_DELAY_HOURS).toStandardDuration()).isAfterNow
+    }
+
+    suspend fun submitIfWanted() {
```
I guess this is the entry point to the whole thing, right?
Can we sleep better if we guard this with a `Mutex` or are we absolutely sure there will be no overlaps? If yes why are we so sure?

kolyaopahle replied:
i am relatively sure because of the delay implied in the scheduled background work, but i will add a mutex to be extra safe :)

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt:
```diff
@@ -5,26 +5,61 @@ import dagger.Provides
+    @Singleton
     @Provides
-    fun newExposureWindows(module: NewExposureWindowsDonor): DonorModule = module
+    fun provideAnalyticsSubmissionApi(
```
Please align with @chris-cwa regarding the http client for the datadonation server, there are some differences in your PRs.

https://github.com/corona-warn-app/cwa-app-android/pull/2311#discussion_r573655003

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt:
```diff
@@ -5,26 +5,61 @@ import dagger.Provides
 class AnalyticsModule {
 
-    @IntoSet
+    @Singleton
```
`Singleton` or `Reusable`, or neither?

BMItter replied:
I'd prefer @Reuseable

kolyaopahle replied:
i agree :)

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsServer.kt:
```diff
@@ -0,0 +1,30 @@
+    suspend fun uploadAnalyticsData(ppaDataRequestAndroid: PpaDataRequestAndroid.PPADataRequestAndroid) {
+        val response = api.get().submitAndroidAnalytics(ppaDataRequestAndroid)
+
+        if (response.code() == 204) {
```
Could I have some logging for each response type/code :relaxed: ?

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsOneTimeWorker.kt:
```diff
@@ -0,0 +1,51 @@
+ *
+ * @see DataDonationAnalyticsScheduler
+ */
+class DataDonationAnalyticsOneTimeWorker @AssistedInject constructor(
```
Do we need a onetime worker? Can win with just the periodic worker?

d4rken replied:
Why do they both look the same?

kolyaopahle replied:
currently the randomness was achieved by scheduling one time work every 24 hours which would then schedule periodic work at a random time in the next 23 hours, but i like the idea of just randomly distributing the initial start delay of the periodic work, will remove the one time worker

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsScheduler.kt:
```diff
@@ -0,0 +1,59 @@
+) {
+
+    /**
+     * Enqueue background deadman notification periodic work
```
deadman?

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsScheduler.kt:
```diff
@@ -0,0 +1,59 @@
+import javax.inject.Inject
+
+/**
+ * We schedule two workers to allow for randomness in the execution times,
```
We just need randomness across the userbase (to be nice to the server), not check on the users device, couldn't we achieve this by setting a random initial delay for the periodic worker?

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt:
```diff
@@ -32,15 +33,24 @@ class OnboardingAnalyticsViewModel @AssistedInject constructor(
+            false
+        }
+
+        launch {
```
If the user smashes that like button, err disable button.... multiple presses can launch this multiple times, right? So donor modules would have to be thread safe on `deleteData`.

kolyaopahle replied:
hmm true, i am also not 100% happy with the positioning here, as we would need to duplicate the code in the AnalyticsSettings view model later on, maybe also some kind of mutex lockout and we move this to a more generic place where it listens to the analyticsEnabled flow preference?

d4rken replied:
Note that FlowPreference is only safe when accessed from the same thread. It's not safe for concurrent access from multiple threads.

I think it makes sense if all deletions, just like starting the analytics submission goes the through `Analytics` main class, and we add a single `Mutex' that guards the deletion and the submission call (as those should not overlap either).

What do you think?

kolyaopahle replied:
sounds good, i've wrapped changes to the property in calls on `Analytics` which should allow us to execute delete on every settings change

d4rken commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:
```diff
@@ -0,0 +1,191 @@
+
+    private fun createDonorModules(): Set<DonorModule> = setOf(exposureRiskMetadataDonor)
+
+    private fun createInstance() = spyk(
```
Why do we need  `spy` here?

kolyaopahle replied:
i was trying to use a spy to check for the return value of the internal check functions in Analytics i.e. `stopDueToProbabilityToSubmit` to not only check that submission was aborted as it is done now, but make sure that it is aborted for the right reason, maybe throwing a more specific exception in the stop functions and raising the exception handling one level further up in the analytics submission could also help with testability

d4rken replied:
Maybe set up the `setup()` that the default state is equal to a successful submission, then in each test case (e.g. probability) change a single parameter and check that the process fails appropriately, maybe by checking that no submission happened.

Or if `stopDueToProbabilityToSubmit` causes an abort, then check that `stopDueToNoUserConsent` was called but `stopDueToLastSubmittedTimestamp`, maybe by verifying calls on the mocks used by those functions.

kolyaopahle replied:
I like it :)

d4rken commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:
```diff
@@ -0,0 +1,191 @@
+            analytics.submitIfWanted()
+        }
+
+        // TODO: is there a way to make sure a spy function returned x?
```
You could inject the randomness source and hardcode it to a specific value.

We currently have `SecureRandom` in the DI graph for SafetyNet.

kolyaopahle replied:
this comment was more in reference to https://github.com/corona-warn-app/cwa-app-android/pull/2280#discussion_r573931976

d4rken commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:
```diff
@@ -0,0 +1,191 @@
+        every { settings.analyticsEnabled } returns mockFlowPreference(false)
+        every { analyticsConfig.probabilityToSubmit } returns 1.0
+        every { settings.lastSubmittedTimestamp } returns mockFlowPreference(
+            Instant.now().minus(Days.TWO.toStandardDuration())
```
Dynamic time in tests is :ghost:

d4rken commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:
```diff
@@ -0,0 +1,191 @@
+
+        coVerify(exactly = 1) {
+            analytics.submitAnalyticsData()
+            dataDonationAnalyticsServer.uploadAnalyticsData(any())
```
Do we have access to the data that would be uploaded such that we can be specific than `any()`?

e.g.

Can we construct `PpaDataRequestAndroid` with the same value? Or would the ProtoBuf equals/hash still not be equal and allow this?

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsApiV1.kt:
```diff
@@ -0,0 +1,13 @@
+
+interface DataDonationAnalyticsApiV1 {
+    @POST("/version/v1/android/dat")
+    suspend fun submitAndroidAnalytics(
```
Can I have a test for this stuff too? :cat2: 

Should be straight forward, check the other API tests: Making sure serialization does what it should, bad data leads to an error, and the API path is what we expect.

d4rken reviewed (commented):
I like what you got

BMItter commented on Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_datadonation.xml:
```diff
@@ -115,6 +115,98 @@
+                    app:layout_constraintTop_toBottomOf="@+id/analytics_last_submit_title"
+                    tools:text="Body" />
+
+            </androidx.constraintlayout.widget.ConstraintLayout>
```
please check for layout Stacking (constraint->constraint->...)
I'm not sure here

kolyaopahle replied:
i'm too stupid to do it without as the stacked layout also contains a card border

BMItter commented on Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt:
```diff
@@ -0,0 +1,87 @@
+                analyticsDir.mkdirs()
+            }
+
+            val dataObject = LastAnalyticsSubmission(
```
Inject and use TimeStamper 👵

BMItter commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt:
```diff
@@ -32,15 +33,24 @@ class OnboardingAnalyticsViewModel @AssistedInject constructor(
+        }
+
         completedOnboardingEvent.postValue(Unit)
-        // TODO: ♫ ♪ Hello Kolya, my old friend ♫ ♪
```
👌 😄

BMItter commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsServer.kt:
```diff
@@ -0,0 +1,30 @@
+    private val api: Lazy<DataDonationAnalyticsApiV1>
+) {
+
+    suspend fun uploadAnalyticsData(ppaDataRequestAndroid: PpaDataRequestAndroid.PPADataRequestAndroid) {
```
consider to use 'when' instead of various if statements if or when you want (optional)

kolyaopahle replied:
done :)

kolyaopahle committed:
Implemented most of the comments from @d4rken

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
if -> when magic

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Moved analytics enable/disable to the Analytics class to allow for deletion on disable

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Made the AnalyticsTest a bit more concise and improved the abort checks

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Changed Api provider to reusable

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Aligned DataDonationAnalyticsApiV1 response with Survery format

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Fixed OnboardingAnalyticsFragmentTest

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'feature/4321-ppa-main' into feature/4823-submit-analytics-data

kolyaopahle committed:
Merge with ppa-main exploded this for some reason, fixed now

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Added analytics deactivation if no config is present on remote

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

BMItter commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt:
```diff
@@ -5,19 +5,66 @@ import de.rki.coronawarnapp.appconfig.AnalyticsConfig
 
 @Reusable
 class AnalyticsConfigMapper @Inject constructor() : AnalyticsConfig.Mapper {
     override fun map(rawConfig: AppConfigAndroid.ApplicationConfigurationAndroid): AnalyticsConfig {
```
using return if... below would be nice

BMItr reviewed (approved):
one optional comment.
lgtm overall 👌

d4rken committed:
Merge branch 'feature/4321-ppa-main' into feature/4823-submit-analytics-data

d4rken committed:
Fix error message if there is no last analytics data.
We only need to check the values if the object itself is non-null.

d4rken committed:
Refactor upload response code check. Use a single `return when` to let the compiler enforce correct case handling.

d4rken committed:
Use a specific time object, as Long and just `getDelay` could be confusing.
Android mostly uses long=miliseconds and not hours.

d4rken committed:
Only get the config once for the whole analytics submission process.
Prevent unnecessary config checks, and if the config cache expires, don't let the config change in the middle of submisison.

d4rken committed:
LINTER ¯\_(ツ)_/¯

harambasicluka commented:
Did on device testing (Pixel 3), logs looked good.

d4rken reviewed (approved):
lgtm
